process/HOPE-3--Dependency-Injection.md

```diff
@@ -0,0 +1,169 @@
+# HOPE 3 -- Dependency Injection
+
+|             |                                             |
+| ------------| ------------------------------------------- |
+| HOPE:       | 3                                           |
+| Title:      | Dependency Injection                        |
+| Author(s):  | Timothy Crosley <timothy.crosley@gmail.com> |
+| Status:     | Proposed                                    |
+| Type:       | Standard Track                              |
+| Created:    | 25-May-2019                                 |
+| Updated:    | 27-May-2019                                 |
+
+## Abstract
+
+This HOPE proposes adding a robust and universally available dependency injection system to Hug.
+This system would be built with the intention of replacing Hug's current `directive` system, one of the first Python microservice dependency injection systems, solving the shortcomings identified upon extensive usage:
+
+- Directives aren't clearly separated from type annotations.
+- It is impossible to have both a directive and type hint for a single parameter.
+- It isn't easy to swap out directives for full HTTP stack testing (through direct calls make it easy to substitute out).
+- Directives aren't nestable.
+
+## Proposed Dependency Injection System
+
+This HOPE proposes a new dependency injection system with an aim for clarity and reuse.
+The API of the new system would consist of a `provides` decorator for defining available dependencies, and a `requires` function for using them:
+
+```python
+@provides(name: str = None, api: hug.API = None, override: bool = False)
+
+def requires(name_or_function: Union[str, Callable], *args, **kwargs) -> Any
```

Comment:
super minor: As you're demonstrating `provides` in the usage form of `@provides`, I'd recommend dropping the `def ` from the `requires` line. Just the `requires(parameters: types) -> returns` seems good.

Reply:
Agreed! Updating :)